### ISSUE TITLE
TemoralMemory.getCellIndices() and ...getCellIndex() may not be necessary

### ISSUE THREAD

**Issue by cogmission** [numenta/nupic#3242]
...and perhaps should not be called?

Please [review here](https://github.com/numenta/nupic/blob/master/src/nupic/research/temporal_memory.py#L849)

They have 2 call sites that don't really need to call them as far as I can tell...

1. https://github.com/numenta/nupic/blob/master/src/nupic/research/temporal_memory.py#L684
2. https://github.com/numenta/nupic/blob/master/src/nupic/research/temporal_memory.py#L706

```Python
  @classmethod
  def getCellIndices(cls, cells):
    """ Returns the indices of the cells passed in.

    @param cells (list) cells to find the indices of
    """
    return [cls.getCellIndex(c) for c in cells]


  @staticmethod
  def getCellIndex(cell):
    """ Returns the index of the cell

    @param cell (int) cell to find the index of
    """
    return cell
```

**Comment by andrewmalta13** [comment]
I was told that these methods were just meant to be exposed to someone using the tm so they don't have to dive into the details about how a cell is represented in the temporal memory. It is the case that there are just integer indices at the moment, but it is possible that a new scheme emerges for performance reasons or otherwise. One could imagine the static method getCellIndex changing and this would not break any code interfacing with the tm.

**Comment by mrcslws** [comment]
Right. With this API, a "cell" is an opaque thing that you shouldn't look at. The proper way to extract the index of the cell is via `tm.getCellIndex(cell)`. It just so happens that today the opaque blob is the cell index.

That's the rationale.

It makes more sense in a dynamic-typed language. In Java it probably feels pretty silly.

I don't have much of an opinion on this. I'd be fine with removing it, and having consumers know that a "cell" is an index.

**Comment by cogmission** [comment]
That's a very interesting consideration. 

Consider this then:

1. If you go with aligning toward the "dynamicness" of the API, then all places where the return type is a list of "cells" should not do operations on them where it is assumed that they are "indexes" and not something else for instance:
```Python
for cell in listOfCells: 
   if cell < someNum or cell == 0
       ...do type violated work here :-P
```

...I suspect there are some cases where the type is assumed to be an integer - but I can't say off-hand now...

I think its cool to go with the "ambience" of the language, but then that style treatment should be uniform across the entire API (not that I know now that it isn't, mind you ;-))  ...but it's up to you guys...

**Comment by rhyolight** [comment]
Closing because the question is answered. If action is required, please create another issue.

**Comment by cogmission** [comment]
Won't this just get shoved on the back burner now and forgotten? To me it requires some next steps:

1. A determination as to the programming convention for the TM - either assuming type opacity or not.
2. If "Opacity" is agreed upon, then a specific review of the API and internals should be done.

Right now we neither have a consensus, nor any impetus to make one - leaving us with a ambiguous API style?

**Comment by rhyolight** [comment]
@cogmission Start a forum thread if you want to discuss it further. This issue is closed because your question is answered.
